nx/lib/nx/binary_backend/matrix.ex

```diff
@@ -7,151 +7,90 @@ defmodule Nx.BinaryBackend.Matrix do
-
-    # if transform_a != none, the upper will become lower and vice-versa
-    lower = transform_a == :none == lower_input
-
-    opts = %{
-      lower: lower,
-      left_side: input_opts[:left_side]
-    }
+    # Here's Paulo Valente's proof in the deleted function ts_handle_opts/3.
+
+    # When lower: false, we need the following procedure
+    # for reusing the lower_triangular engine:
+    #
+    # First, we need to reverse both rows and columns
+    # so we can turn an upper-triangular matrix
+    # into a lower-triangular one.
+    # The result will also be reversed in this case.
+    #
+    # Proof:
+    # For a result [x1, x2, x3.., xn] and a row [a1, a2, a4, ..., an]
+    # we have the corresponding b = a1 * x1 + a2 * x2 + a3 * x3 + ...+ an * xn
+    # Since the addition of a_i * x_i is commutative, by reversing the columns
+    # of a, the yielded x will be reversed.
+    # Furthermore, if we reverse the rows of a, we need to reverse the rows of b
+    # so each row is kept together with it's corresponding result.
+    #
+    # For example, the system:
+    # A = [[a b c], [0 d e], [0 0 f]]
+    # b = [b1, b2, b3, b4]
+    # which yields x = [x1, x2, x3, x4]
+    # is therefore equivalent to:
+    # A = [[f 0 0], [e d 0], [c b a]]
+    # b = [b4, b3, b2, b1]
+    # which yields [x4, x3, x2, x1]
+
+    # For handling left_side: false
+    # Let's notate the system matrix as L when it's a lower triangular matrix
+    # and U when it's upper triangular
+    #
+    # To solve X.L = B, we can then transpose both sides:
+    # transpose(X.L) = transpose(L).X_t = U.X_t = b_t
+    # This equation, in turn, has the same shape (U.X = B) as the one we can solve through
+    # applying `ts_handle_lower_opt` properly, which would yield X_t.
+    # Transposing the result suffices for yielding the final result.
+
+    opt_left_side = !!input_opts[:left_side]
+    opt_lower = !!input_opts[:lower]
+    opt_transpose_a = (input_opts[:transform_a] == :transpose)
+
+    transpose_a? = opt_left_side == opt_transpose_a
+    reverse_a? = opt_left_side == (opt_lower == opt_transpose_a)
+
+    transpose_b? = (not opt_left_side) and match?({_, _}, b_shape)
+    reverse_b? = reverse_a?
+
+    transpose_x? = transpose_b?
+    reverse_x? = reverse_b?
 
     a_matrix =
       a_data
       |> binary_to_matrix(a_type, a_shape)
-      |> ts_transform_a(transform_a)
-      |> ts_handle_opts(opts, :a)
+      |> then(& transpose_a? && transpose_matrix(&1) || &1)
+      |> then(& reverse_a? && reverse_matrix(&1) || &1)
 
     b_matrix_or_vec =
       case b_shape do
         {_rows, _cols} ->
-          b_data |> binary_to_matrix(b_type, b_shape) |> ts_handle_opts(opts, :b)
+          b_data |> binary_to_matrix(b_type, b_shape)
 
         {_rows} ->
-          b_data |> binary_to_vector(b_type) |> ts_handle_opts(opts, :b)
+          b_data |> binary_to_vector(b_type)
+      end
+      |> then(& transpose_b? && transpose_matrix(&1) || &1)
```

Comment:
```suggestion
      |> then(&if(transpose_b?, do: transpose_matrix(&1), else: &1))
```

Stylistically, I find inline ifs instead of using && and || for a ternary selection to be a more clear option because it doesn't require thinking about operator precedence